# Exercise 02 -  OLAP Cubes - Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by creating and connecting to the database by running the cells below.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Start with a simple cube
TODO: Write a query that calculates the revenue (sales_amount) by day, rating, and city. Remember to join with the appropriate dimension tables to replace the keys with the dimension labels. Sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below.

In [10]:
%%time
%%sql

SELECT SUM(sales_amount) as rev, rating, day, city
FROM factsales fa
JOIN dimmovie dim ON (dim.movie_key = fa.movie_key)
JOIN dimdate da ON (da.date_key = fa.date_key)
JOIN dimcustomer dc ON (dc.customer_key = fa.customer_key)
GROUP BY day, rating, city
ORDER BY rev desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.97 ms, sys: 185 µs, total: 5.15 ms
Wall time: 36 ms


[(Decimal('24.97'), 'G', 30, 'San Bernardino'),
 (Decimal('23.95'), 'NC-17', 30, 'Apeldoorn'),
 (Decimal('22.97'), 'NC-17', 21, 'Belm'),
 (Decimal('21.97'), 'R', 28, 'Mwanza'),
 (Decimal('21.97'), 'PG-13', 30, 'Zanzibar'),
 (Decimal('21.97'), 'G', 21, 'Citt del Vaticano'),
 (Decimal('19.97'), 'R', 22, 'Yangor'),
 (Decimal('19.97'), 'R', 1, 'Qomsheh'),
 (Decimal('19.97'), 'G', 17, 'Rajkot'),
 (Decimal('19.97'), 'PG-13', 28, 'Dhaka'),
 (Decimal('19.96'), 'PG', 19, 'Najafabad'),
 (Decimal('19.95'), 'R', 30, 'Fengshan'),
 (Decimal('18.98'), 'PG', 28, 'So Leopoldo'),
 (Decimal('18.98'), 'G', 21, 'Wroclaw'),
 (Decimal('18.97'), 'NC-17', 1, 'Memphis'),
 (Decimal('18.97'), 'G', 30, 'Omdurman'),
 (Decimal('18.97'), 'PG-13', 29, 'Shimoga'),
 (Decimal('18.97'), 'PG-13', 30, 'Osmaniye'),
 (Decimal('18.97'), 'PG', 19, 'Sokoto'),
 (Decimal('18.95'), 'PG-13', 21, 'Asuncin')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

TODO: Write a query that reduces the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below. 

In [11]:
%%time
%%sql

SELECT SUM(sales_amount) as rev, rating, day, city
FROM factsales fa
JOIN dimmovie dim ON (dim.movie_key = fa.movie_key)
JOIN dimdate da ON (da.date_key = fa.date_key)
JOIN dimcustomer dc ON (dc.customer_key = fa.customer_key)
WHERE rating='PG-13'
GROUP BY day, rating, city
ORDER BY rev desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 1.62 ms, sys: 3.54 ms, total: 5.17 ms
Wall time: 16.4 ms


[(Decimal('21.97'), 'PG-13', 30, 'Zanzibar'),
 (Decimal('19.97'), 'PG-13', 28, 'Dhaka'),
 (Decimal('18.97'), 'PG-13', 30, 'Osmaniye'),
 (Decimal('18.97'), 'PG-13', 29, 'Shimoga'),
 (Decimal('18.95'), 'PG-13', 21, 'Asuncin'),
 (Decimal('17.98'), 'PG-13', 21, 'Parbhani'),
 (Decimal('17.98'), 'PG-13', 20, 'Baha Blanca'),
 (Decimal('17.98'), 'PG-13', 30, 'Nagareyama'),
 (Decimal('17.96'), 'PG-13', 30, 'Tanauan'),
 (Decimal('17.95'), 'PG-13', 17, 'Ikerre'),
 (Decimal('16.98'), 'PG-13', 30, 'Zhoushan'),
 (Decimal('16.97'), 'PG-13', 30, 'Yerevan'),
 (Decimal('16.97'), 'PG-13', 30, 'Newcastle'),
 (Decimal('15.98'), 'PG-13', 20, 'Santa Rosa'),
 (Decimal('15.98'), 'PG-13', 17, 'Yantai'),
 (Decimal('15.98'), 'PG-13', 30, 'Santa Rosa'),
 (Decimal('15.98'), 'PG-13', 8, 'Pontianak'),
 (Decimal('15.97'), 'PG-13', 30, 'Toulouse'),
 (Decimal('15.97'), 'PG-13', 23, 'Tiefa'),
 (Decimal('15.97'), 'PG-13', 1, 'Moscow')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

TODO: Write a query to create a subcube of the initial cube that includes moves with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

The first few rows of your output should match the table below. 

In [12]:
%%time
%%sql

SELECT SUM(sales_amount) as rev, rating, day, city
FROM factsales fa
JOIN dimmovie dim ON (dim.movie_key = fa.movie_key)
JOIN dimdate da ON (da.date_key = fa.date_key)
JOIN dimcustomer dc ON (dc.customer_key = fa.customer_key)
WHERE (rating='PG-13' or rating='PG') and (city='Bellevue' or city='Lancaster') and (day=1 or day=15 or day=30)
GROUP BY day, rating, city
ORDER BY rev desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
6 rows affected.
CPU times: user 6.24 ms, sys: 437 µs, total: 6.67 ms
Wall time: 11.2 ms


[(Decimal('12.98'), 'PG', 30, 'Lancaster'),
 (Decimal('5.99'), 'PG-13', 1, 'Lancaster'),
 (Decimal('3.99'), 'PG-13', 30, 'Bellevue'),
 (Decimal('2.99'), 'PG-13', 30, 'Lancaster'),
 (Decimal('1.98'), 'PG-13', 15, 'Bellevue'),
 (Decimal('0.99'), 'PG', 1, 'Bellevue')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>